### ISSUE TITLE
Add the "test" classpath attribute to test sources and dependencies

### ISSUE THREAD

**Issue by fbricon** [eclipse/buildship#689]
Eclipse Photon now support separation of main and test sources.
Classpath entries need a `test` attribute like:
```xml
<classpathentry kind="src" output="target/test-classes" path="src/test/java">
  <attributes>
    <attribute name="test" value="true"/>
  </attributes>
</classpathentry>
```
Same goes for binary entries.

See https://bugs.eclipse.org/bugs/show_bug.cgi?id=224708 for more information. @brychcy should be able to give more details if needed.

**Comment by fbricon** [comment]
See how Maven projects look like with those changes:
<img width="315" alt="test-entries" src="https://user-images.githubusercontent.com/148698/38526960-6f7dffa8-3c26-11e8-992b-b9afd0491b43.png">

**Comment by donat** [comment]
We'll definitely want to look into this. Thank you for the link!

**Comment by lukeu** [comment]
Presumably adding the 'test' attribute to sources would be done in Gradle rather than Buildship? [Gradle issue 4802](https://github.com/gradle/gradle/issues/4802) appears to exist for that.

**Comment by fbricon** [comment]
@lukeu that would only apply for sources. You also need Buildship to add the attribute to the matching classpath entries of the Gradle classpath container. This can't be expressed in the .classpath file

**Comment by lukeu** [comment]
@fbricon Right, I see now - the generated classpath(es) need to be flagged too.

Incidentally I had a tinker in the source, mainly for my own curiosity. I got the Gradle bit working (just a hack, not a complete solution). However I hit a blocker with Buildship (installed as per [the dev instructions](https://github.com/eclipse/buildship/blob/master/docs/development/Setup.md)) in that I couldn't figure out how to make it work with Photon, meaning I couldn't test it. Oomph gave some conflict error.

**Comment by lukeu** [comment]
> However I hit a blocker with Buildship (installed as per the dev instructions) in that I couldn't figure out how to make it work with Photon

Just to say I'm now no longer blocked. I...
 * Built gradle from source, fixed [a crash](https://github.com/gradle/gradle/issues/5278)
 * Used the old dev instructions
 * Copied `tooling-e47.target` and cargo-crated it around to reference Eclipse's Photon p2 repo

and now have my own builds of Gradle and Buildship running with Photon. Yay! :-)

Note however: it's not really my intent to fix this, I'm mainly learning my way around for educational/interest sake. I think there's too many factors to take into consideration (I know nothing about Java 9 modules, for example) to feel I can have good test coverage of other people's configurations. I will probably just fix a few other easy tickets of interest to me.

Still, I at least have have test source dirs & runtime libraries paths flagged as "test" for my a small test workspace. The next step would be to support for other "type Test" tasks, like integration tests. I'm also curious under what circumstances I would need to set `IClasspathAttribute.WITHOUT_TEST_CODE` (when projects wish to make their test-code private to external modules?) I can share any of these small code snippets if they are of interest to anyone.

**Comment by donat** [comment]
@lukeu Regarding the development environment, I've just added a `tooling-e48.target` file so that you can develop Buildship against Eclipse Photon. Also, you can execute the headless build with the `./gradlew clean build -Peclipse.version=48` command.

**Comment by lukeu** [comment]
@donat, thanks that should make life easier

**Comment by howlger** [comment]
Workaround that works for me in Eclipse Photon RC2: https://github.com/gradle/gradle/issues/4802#issuecomment-396165166

**Comment by learnerEvermore** [comment]
This is of great importance and interest. It would partially resolve [Main vs. test, compile vs runtime classpaths in Eclipse - once and for all, how?](https://discuss.gradle.org/t/main-vs-test-compile-vs-runtime-classpaths-in-eclipse-once-and-for-all-how/17403)

**Comment by ChristianCiach** [comment]
> Workaround that works for me in Eclipse Photon RC2: [gradle/gradle#4802 (comment)](https://github.com/gradle/gradle/issues/4802#issuecomment-396165166)

Unfortunately this doesn't fully work with Buildship. The workaround sets the "test" attribute for the libraries, too, but this doesn't work with Buildship, because with Buildship the classpath doesn't list the individual libraries, but only the "buildship classpath container" that cannot be customized by any workarounds. 

I think the cleanest solution would be for Buildship to use different classpath containers for the main and test source sets. In this case, the *test* classpath container could be marked as *test* as a whole.

**Comment by oehme** [comment]
> but this doesn't work with Buildship, because with Buildship the classpath doesn't list the individual libraries, but only the "buildship classpath container" that cannot be customized by any workarounds.

That's not true. The classpath container contains exactly the libraries you configured in your `eclipse {}` model, including any attributes you configured on them.

**Comment by donat** [comment]
FYI I've implemented the changes in Gradle: https://github.com/gradle/gradle/pull/9484 (waiting for review). Most probably the PR will be merged for Gradle 5.6.

**Comment by donat** [comment]
A little update: we have implemented the necessary changes in #905. The upcoming Buildship release will make use of the test sources feature (with Gradle 5.6+)
